Fill out tutorial on unneeded mutations (#177)
tskit-dev/tutorials: no_mutations.md

  name: python3
---

```{currentmodule} tskit
```

(sec_tskit_no_mutations)=

# Do you really need mutations?

> In tree sequences, the genetic genealogy exists independently of the mutations that

Reviewer: ```suggestion
In tree sequences, the genealogy exists independently of the mutations that
```

  name: python3
---

```{currentmodule} tskit
```

(sec_tskit_no_mutations)=

# Do you really need mutations?

> In tree sequences, the genetic genealogy exists independently of the mutations that
> generate genetic variation. This tutorial aims to illustrate
> cases where the genealogy on its own may be sufficent; a number of
> analyses do not actually require genetic variation.

Reviewer: ```suggestion
In tree sequences, the genetic genealogy exists independently of the mutations that
generate genetic variation, and often we are primarily interested in genetic variation
because of what it can tell us about those genealogies. This tutorial aims to illustrate
when we can leave mutations and genetic variation aside and study the genealogies directly.
```

Author: Nice!

> This applies particularly to cases where you generate tree sequences by simulation.
> In this case, although simulating variable sites allows you to generate actual genetic
> data, it's worth considering the following points:

Reviewer: ```suggestion
In simulations we know the true genealogies, and so it can be very helpful to work
with these directly. 
In real data, we might infer the trees and then work with the resulting genealogies.
(Of course, mutations add additional noise, and would be necessary
to produce data directly comparable to sequencing data.)
If you're wondering whether you need to add mutations at all,
it's worth considering the following points:
```

1. Neutral mutations and sites can be added to a genealogy later, if you find they *are*
needed

> 2. Simulating sites and mutations increases memory requirements and takes marginally
> more CPU time
> 3. Storing mutations increases the tree sequence file size

Reviewer: ```suggestion
2. Simulating sites and mutations increases memory requirements and tree sequence file size somewhat
```
why not just leave it at this? Or, could add "and CPU time (although usually this is inconsequential)"

Author: I think the CPU time thing might be worth adding because of e.g. https://github.com/tskit-dev/msprime/discussions/1687

Reviewer: Yes, I saw that, and was trying to write it in, but the sentence wasn't being very natural and leaving out CPU doesn't imply that CPU is never a problem. I vote against leaving in "takes marginally more CPU time" since that isn't a good reason; I think the good reason is that "in some cases it takes significantly more CPU time", but that was seeming too long. Up to you?

Reviewer: never mind, I see you put in something to this effect.

> 4. Some genetic analyses can be performed equally well (or better!) on tree sequences
> that have no sites or mutations.

Reviewer: ```suggestion
3. Quantities of interest can often be inferred equally well (or better!) on tree sequences
that have no sites or mutations.
```
Changing this because saying "genetic analyses" really sounds like you're using real data, where you definitely do have mutations.

Author: Good point

    ploidy=2,
    recombination_rate=rho,
    sequence_length=L,
    random_seed=123,
)

# Optionally, add neutral mutations later, after simulating. This takes some CPU time
# (although it is usually fast compared to simulating the original tree sequence)
ts_mutated = msprime.sim_mutations(ts_no_mut, rate=mu, random_seed=456)
print(
    "Adding mutations has increased the tree sequence file size by "
    f"{ts_mutated.nbytes / ts_no_mut.nbytes * 100:.0f}%",
)
```

:::{note}
Above we have overlaid sites and mutations onto an existing tree sequence by simulation.
It is also possible to overlay mutations in such a way as to generate a known pattern of
genetic variation, using {meth}`tskit.Tree.map_mutations`.
:::

Although including mutations has increased the file size a fair bit in this example,
unless you are running simulations with vast numbers of mutations, tree sequences are
usually compact enough that the cost of including them shouldn't be prohibitive.
Nevertheless, it may not be obvious that it is perfectly possible, indeed sometimes
preferable, to perform genetic analyses on tree sequences that do not contain mutations

> and variable sites. This is the focus of the remainder of this tutorial.

Reviewer: At first I was averse to saying "genetic analyses" here - it felt weird - but maybe that's the point, and I think it's ok?

    msprime.sim_mutations(ts, rate=mu/100, random_seed=i+4) 
    for i, ts in enumerate(ts_reps)
]

# Return sample sets as all samples from each population (uses all pairwise comparisons)
def sample_sets(ts):
    return [ts.samples(population=p.id) for p in ts.populations()]

Fst_genealogy = np.array([ts.Fst(sample_sets(ts), mode="branch") for ts in ts_reps])
Fst_genetic_var = np.array([ts.Fst(sample_sets(ts)) for ts in ts_mutated_reps])

# For the theoretical expectation see e.g. Crow and Aoki (1984) PNAS 81: 6073, eqn 7
Fst_theory = 1/(4*subpop_size*migration_rate*(n_subpops/(n_subpops-1))**2 + 1)

plt.scatter(["Genetic variation"] * 20, Fst_genetic_var)
plt.scatter(["Genealogy"] * 20, Fst_genealogy)
plt.xlabel("Basis of estimate")
plt.ylabel("Fst\n(20 replicates)")

plt.axhline(y=Fst_theory, ls=":", c="grey")
plt.text(0.5, Fst_theory, 'theoretical prediction', ha='center')
plt.show()
```

Therefore, if your ultimate goal is to compare or estimate genetic statistics of this
sort (rather than to examine the genetic sequence itself), then using the
genealogy-based approach should give you more statistical power.

> 

Reviewer: ```suggestion
(However, in real data there is additional noise introduced by
inference of the underlying tree sequence which is not accounted for below.)

```

the genome are small, each window will contain relatively few mutations, and statistics
based on the genetic variation generated by mutations will be be subject to greater
mutational noise than those based on branch lengths in the genealogy. Here's an
example using the basic genetic {meth}`~tskit.TreeSequence.diversity`
in 1Kb windows along our simulated genome:

```{code-cell} ipython3
fig, (ax1, ax2) = plt.subplots(ncols=2, figsize=(12, 3), sharey=True)
w = np.linspace(0, L, num=L//1_000)
ax1.stairs(ts_mutated.diversity(windows=w), w/1_000, baseline=None)
ax1.set_ylabel("Diversity")
ax1.set_xlabel("Genome position (kb)")
ax1.set_title("Site-based calculation")
ax1.set_yscale("log")
ax2.stairs(ts_no_mut.diversity(windows=w, mode="branch") * mu, w/1_000, baseline=None)
ax2.set_xlabel("Genome position (kb)")
ax2.set_title("Branch-length-based calculation")
ax2.set_yscale("log")
plt.show()
```

## Summary

In summary, if you are analysing tree sequences, especially those produced by simulation,
think carefully. Do you really need to analyse information about alleles and mutations?
If not, you may be able to omit sites and mutations from the tree sequence,

> yet still retain the ability to calculate parameters of interest. This genealogical
> approach can even produce better statistical estimates than those explicitly based
> on genetic variation.

Reviewer: ```suggestion
yet still retain the ability to calculate parameters of interest. This genealogical
approach can provide more accurate descriptors of the quantities of interest
than those explicitly based on genetic variation.
```
